src/test/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcherSpec.scala

```diff
@@ -124,6 +129,95 @@ class BlockFetcherSpec extends TestKit(ActorSystem("BlockFetcherSpec_System")) w
       peersClient.expectMsgClass(classOf[BlacklistPeer])
       peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == firstGetBlockHeadersRequest => () }
     }
+
+    "should not append new blocks if the received data does not match" in new TestSetup {
+
+      // Important: Here we are forcing the mismatch between request headers and received bodies
+      override lazy val validators = new MockValidatorsFailingOnBlockBodies
+
+      startFetcher()
+
+      val getBlockHeadersRequest =
+        GetBlockHeaders(Left(1), syncConfig.blockHeadersPerRequest, skip = 0, reverse = false)
+      peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == getBlockHeadersRequest => () }
+
+      val chain = BlockHelpers.generateChain(syncConfig.blockHeadersPerRequest, FixtureBlocks.Genesis.block)
+      val getBlockHeadersResponse = BlockHeaders(chain.map(_.header))
+      peersClient.reply(PeersClient.Response(fakePeer, getBlockHeadersResponse))
+
+      val getBlockBodiesRequest = GetBlockBodies(chain.map(_.hash))
+      peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == getBlockBodiesRequest => () }
+
+      // This response will be invalid given we are using a special validator!
+      val getBlockBodiesResponse = BlockBodies(chain.map(_.body))
+      peersClient.reply(PeersClient.Response(fakePeer, getBlockBodiesResponse))
+
+      peersClient.expectMsgClass(classOf[BlacklistPeer])
+      peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == getBlockBodiesRequest => () }
+    }
+
+    "should be able to handle block bodies received in several parts" in new TestSetup {
+
+      startFetcher()
+
+      val getBlockHeadersRequest =
+        GetBlockHeaders(Left(1), syncConfig.blockHeadersPerRequest, skip = 0, reverse = false)
+      peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == getBlockHeadersRequest => () }
+
+      val chain = BlockHelpers.generateChain(syncConfig.blockHeadersPerRequest, FixtureBlocks.Genesis.block)
+
+      val getBlockHeadersResponse = BlockHeaders(chain.map(_.header))
+      peersClient.reply(PeersClient.Response(fakePeer, getBlockHeadersResponse))
+
+      val getBlockBodiesRequest1 = GetBlockBodies(chain.map(_.hash))
+      peersClient.expectMsgPF() { case PeersClient.Request(msg, _, _) if msg == getBlockBodiesRequest1 => () }
+
+      // It will receive all the requested bodies, but splitted in 2 parts.
+      val (subChain1, subChain2) = chain.splitAt(syncConfig.blockHeadersPerRequest / 2)
```

Comment:
block bodies per request? I know these are same values in tests, but the right config value should be used anyway I think